{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cffabf30",
   "metadata": {},
   "source": [
    "Dataset link\n",
    "\n",
    "https://chanzuckerberg.github.io/cellxgene-census/notebooks/api_demo/census_datasets.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b4a5fa39",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python 3.10.18\n"
     ]
    }
   ],
   "source": [
    "!python --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "38dcd630",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>citation</th>\n",
       "      <th>collection_id</th>\n",
       "      <th>collection_name</th>\n",
       "      <th>collection_doi</th>\n",
       "      <th>collection_doi_label</th>\n",
       "      <th>dataset_id</th>\n",
       "      <th>dataset_version_id</th>\n",
       "      <th>dataset_title</th>\n",
       "      <th>dataset_h5ad_path</th>\n",
       "      <th>dataset_total_cell_count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>soma_joinid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Publication: https://doi.org/10.1016/j.isci.20...</td>\n",
       "      <td>8e880741-bf9a-4c8e-9227-934204631d2a</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>10.1016/j.isci.2022.104097</td>\n",
       "      <td>Marshall et al. (2022) iScience</td>\n",
       "      <td>4eb29386-de81-452f-b3c0-e00844e8c7fd</td>\n",
       "      <td>f76861bb-becb-4eb7-82fc-782dc96ccc7f</td>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191112_05</td>\n",
       "      <td>4eb29386-de81-452f-b3c0-e00844e8c7fd.h5ad</td>\n",
       "      <td>10888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Publication: https://doi.org/10.1016/j.isci.20...</td>\n",
       "      <td>8e880741-bf9a-4c8e-9227-934204631d2a</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>10.1016/j.isci.2022.104097</td>\n",
       "      <td>Marshall et al. (2022) iScience</td>\n",
       "      <td>78d59e4a-82eb-4a61-a1dc-da974d7ea54b</td>\n",
       "      <td>7d7ec1b6-6e3f-4aaa-9442-4b22f3424396</td>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191112_08</td>\n",
       "      <td>78d59e4a-82eb-4a61-a1dc-da974d7ea54b.h5ad</td>\n",
       "      <td>10250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Publication: https://doi.org/10.1016/j.isci.20...</td>\n",
       "      <td>8e880741-bf9a-4c8e-9227-934204631d2a</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>10.1016/j.isci.2022.104097</td>\n",
       "      <td>Marshall et al. (2022) iScience</td>\n",
       "      <td>add5eb84-5fc9-4f01-982e-a346dd42ee82</td>\n",
       "      <td>de54aed8-4f73-48f6-9229-418a840e2d82</td>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191109_20</td>\n",
       "      <td>add5eb84-5fc9-4f01-982e-a346dd42ee82.h5ad</td>\n",
       "      <td>12906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Publication: https://doi.org/10.1016/j.isci.20...</td>\n",
       "      <td>8e880741-bf9a-4c8e-9227-934204631d2a</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>10.1016/j.isci.2022.104097</td>\n",
       "      <td>Marshall et al. (2022) iScience</td>\n",
       "      <td>b020294c-ab82-4547-b5a7-63d8ffa575ed</td>\n",
       "      <td>abe4fce1-0859-4a56-ad1e-734d79f0e6c8</td>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191112_13</td>\n",
       "      <td>b020294c-ab82-4547-b5a7-63d8ffa575ed.h5ad</td>\n",
       "      <td>15161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Publication: https://doi.org/10.1038/s41591-02...</td>\n",
       "      <td>a96133de-e951-4e2d-ace6-59db8b3bfb1d</td>\n",
       "      <td>HTAN/HTAPP Broad - Spatio-molecular dissection...</td>\n",
       "      <td>10.1038/s41591-024-03215-z</td>\n",
       "      <td>Klughammer et al. (2024) Nat Med</td>\n",
       "      <td>d7476ae2-e320-4703-8304-da5c42627e71</td>\n",
       "      <td>863fc5e4-bd4a-4681-9c3d-0ee7ef54e327</td>\n",
       "      <td>HTAPP-330-SMP-1082 scRNA-seq</td>\n",
       "      <td>d7476ae2-e320-4703-8304-da5c42627e71.h5ad</td>\n",
       "      <td>565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1568</th>\n",
       "      <td>Publication: https://doi.org/10.1038/s41586-02...</td>\n",
       "      <td>45d5d2c3-bc28-4814-aed6-0bb6f0e11c82</td>\n",
       "      <td>A single-cell transcriptional timelapse of mou...</td>\n",
       "      <td>10.1038/s41586-024-07069-w</td>\n",
       "      <td>Qiu et al. (2024) Nature</td>\n",
       "      <td>dcfa2614-7ca7-4d82-814c-350626eccb26</td>\n",
       "      <td>3002a659-a1a9-4406-9976-99e658e1fbb5</td>\n",
       "      <td>Major cell cluster: Mesoderm</td>\n",
       "      <td>dcfa2614-7ca7-4d82-814c-350626eccb26.h5ad</td>\n",
       "      <td>3267338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1569</th>\n",
       "      <td>Publication: https://doi.org/10.1126/science.a...</td>\n",
       "      <td>e5f58829-1a66-40b5-a624-9046778e74f5</td>\n",
       "      <td>Tabula Sapiens</td>\n",
       "      <td>10.1126/science.abl4896</td>\n",
       "      <td>The Tabula Sapiens Consortium* et al. (2022) S...</td>\n",
       "      <td>53d208b0-2cfd-4366-9866-c3c6114081bc</td>\n",
       "      <td>10df7690-6d10-4029-a47e-0f071bb2df83</td>\n",
       "      <td>Tabula Sapiens - All Cells</td>\n",
       "      <td>53d208b0-2cfd-4366-9866-c3c6114081bc.h5ad</td>\n",
       "      <td>1136218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1570</th>\n",
       "      <td>Publication: https://doi.org/10.1038/s41586-02...</td>\n",
       "      <td>45d5d2c3-bc28-4814-aed6-0bb6f0e11c82</td>\n",
       "      <td>A single-cell transcriptional timelapse of mou...</td>\n",
       "      <td>10.1038/s41586-024-07069-w</td>\n",
       "      <td>Qiu et al. (2024) Nature</td>\n",
       "      <td>dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3</td>\n",
       "      <td>3817734b-0f82-433b-8c38-55b214200fff</td>\n",
       "      <td>Whole dataset: Raw counts only</td>\n",
       "      <td>dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3.h5ad</td>\n",
       "      <td>11441407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1571</th>\n",
       "      <td>Publication: https://doi.org/10.1038/s41593-02...</td>\n",
       "      <td>1ca90a2d-2943-483d-b678-b809bf464c30</td>\n",
       "      <td>SEA-AD: Seattle Alzheimer’s Disease Brain Cell...</td>\n",
       "      <td>10.1038/s41593-024-01774-5</td>\n",
       "      <td>Gabitto et al. (2024) Nat Neurosci</td>\n",
       "      <td>c2876b1b-06d8-4d96-a56b-5304f815b99a</td>\n",
       "      <td>c32964d2-3339-441f-8e56-7177234c7876</td>\n",
       "      <td>Whole Taxonomy - MTG: Seattle Alzheimer's Dise...</td>\n",
       "      <td>c2876b1b-06d8-4d96-a56b-5304f815b99a.h5ad</td>\n",
       "      <td>1226855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1572</th>\n",
       "      <td>Publication: https://doi.org/10.1038/s41593-02...</td>\n",
       "      <td>1ca90a2d-2943-483d-b678-b809bf464c30</td>\n",
       "      <td>SEA-AD: Seattle Alzheimer’s Disease Brain Cell...</td>\n",
       "      <td>10.1038/s41593-024-01774-5</td>\n",
       "      <td>Gabitto et al. (2024) Nat Neurosci</td>\n",
       "      <td>6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3</td>\n",
       "      <td>d3427e8c-c55d-4d4e-b15b-1a8774cd3a4b</td>\n",
       "      <td>Whole Taxonomy - DLPFC: Seattle Alzheimer's Di...</td>\n",
       "      <td>6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3.h5ad</td>\n",
       "      <td>1309414</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1573 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      citation  \\\n",
       "soma_joinid                                                      \n",
       "0            Publication: https://doi.org/10.1016/j.isci.20...   \n",
       "1            Publication: https://doi.org/10.1016/j.isci.20...   \n",
       "2            Publication: https://doi.org/10.1016/j.isci.20...   \n",
       "3            Publication: https://doi.org/10.1016/j.isci.20...   \n",
       "4            Publication: https://doi.org/10.1038/s41591-02...   \n",
       "...                                                        ...   \n",
       "1568         Publication: https://doi.org/10.1038/s41586-02...   \n",
       "1569         Publication: https://doi.org/10.1126/science.a...   \n",
       "1570         Publication: https://doi.org/10.1038/s41586-02...   \n",
       "1571         Publication: https://doi.org/10.1038/s41593-02...   \n",
       "1572         Publication: https://doi.org/10.1038/s41593-02...   \n",
       "\n",
       "                                    collection_id  \\\n",
       "soma_joinid                                         \n",
       "0            8e880741-bf9a-4c8e-9227-934204631d2a   \n",
       "1            8e880741-bf9a-4c8e-9227-934204631d2a   \n",
       "2            8e880741-bf9a-4c8e-9227-934204631d2a   \n",
       "3            8e880741-bf9a-4c8e-9227-934204631d2a   \n",
       "4            a96133de-e951-4e2d-ace6-59db8b3bfb1d   \n",
       "...                                           ...   \n",
       "1568         45d5d2c3-bc28-4814-aed6-0bb6f0e11c82   \n",
       "1569         e5f58829-1a66-40b5-a624-9046778e74f5   \n",
       "1570         45d5d2c3-bc28-4814-aed6-0bb6f0e11c82   \n",
       "1571         1ca90a2d-2943-483d-b678-b809bf464c30   \n",
       "1572         1ca90a2d-2943-483d-b678-b809bf464c30   \n",
       "\n",
       "                                               collection_name  \\\n",
       "soma_joinid                                                      \n",
       "0            High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "1            High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "2            High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "3            High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "4            HTAN/HTAPP Broad - Spatio-molecular dissection...   \n",
       "...                                                        ...   \n",
       "1568         A single-cell transcriptional timelapse of mou...   \n",
       "1569                                            Tabula Sapiens   \n",
       "1570         A single-cell transcriptional timelapse of mou...   \n",
       "1571         SEA-AD: Seattle Alzheimer’s Disease Brain Cell...   \n",
       "1572         SEA-AD: Seattle Alzheimer’s Disease Brain Cell...   \n",
       "\n",
       "                         collection_doi  \\\n",
       "soma_joinid                               \n",
       "0            10.1016/j.isci.2022.104097   \n",
       "1            10.1016/j.isci.2022.104097   \n",
       "2            10.1016/j.isci.2022.104097   \n",
       "3            10.1016/j.isci.2022.104097   \n",
       "4            10.1038/s41591-024-03215-z   \n",
       "...                                 ...   \n",
       "1568         10.1038/s41586-024-07069-w   \n",
       "1569            10.1126/science.abl4896   \n",
       "1570         10.1038/s41586-024-07069-w   \n",
       "1571         10.1038/s41593-024-01774-5   \n",
       "1572         10.1038/s41593-024-01774-5   \n",
       "\n",
       "                                          collection_doi_label  \\\n",
       "soma_joinid                                                      \n",
       "0                              Marshall et al. (2022) iScience   \n",
       "1                              Marshall et al. (2022) iScience   \n",
       "2                              Marshall et al. (2022) iScience   \n",
       "3                              Marshall et al. (2022) iScience   \n",
       "4                             Klughammer et al. (2024) Nat Med   \n",
       "...                                                        ...   \n",
       "1568                                  Qiu et al. (2024) Nature   \n",
       "1569         The Tabula Sapiens Consortium* et al. (2022) S...   \n",
       "1570                                  Qiu et al. (2024) Nature   \n",
       "1571                        Gabitto et al. (2024) Nat Neurosci   \n",
       "1572                        Gabitto et al. (2024) Nat Neurosci   \n",
       "\n",
       "                                       dataset_id  \\\n",
       "soma_joinid                                         \n",
       "0            4eb29386-de81-452f-b3c0-e00844e8c7fd   \n",
       "1            78d59e4a-82eb-4a61-a1dc-da974d7ea54b   \n",
       "2            add5eb84-5fc9-4f01-982e-a346dd42ee82   \n",
       "3            b020294c-ab82-4547-b5a7-63d8ffa575ed   \n",
       "4            d7476ae2-e320-4703-8304-da5c42627e71   \n",
       "...                                           ...   \n",
       "1568         dcfa2614-7ca7-4d82-814c-350626eccb26   \n",
       "1569         53d208b0-2cfd-4366-9866-c3c6114081bc   \n",
       "1570         dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3   \n",
       "1571         c2876b1b-06d8-4d96-a56b-5304f815b99a   \n",
       "1572         6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3   \n",
       "\n",
       "                               dataset_version_id  \\\n",
       "soma_joinid                                         \n",
       "0            f76861bb-becb-4eb7-82fc-782dc96ccc7f   \n",
       "1            7d7ec1b6-6e3f-4aaa-9442-4b22f3424396   \n",
       "2            de54aed8-4f73-48f6-9229-418a840e2d82   \n",
       "3            abe4fce1-0859-4a56-ad1e-734d79f0e6c8   \n",
       "4            863fc5e4-bd4a-4681-9c3d-0ee7ef54e327   \n",
       "...                                           ...   \n",
       "1568         3002a659-a1a9-4406-9976-99e658e1fbb5   \n",
       "1569         10df7690-6d10-4029-a47e-0f071bb2df83   \n",
       "1570         3817734b-0f82-433b-8c38-55b214200fff   \n",
       "1571         c32964d2-3339-441f-8e56-7177234c7876   \n",
       "1572         d3427e8c-c55d-4d4e-b15b-1a8774cd3a4b   \n",
       "\n",
       "                                                 dataset_title  \\\n",
       "soma_joinid                                                      \n",
       "0             Spatial transcriptomics in mouse: Puck_191112_05   \n",
       "1             Spatial transcriptomics in mouse: Puck_191112_08   \n",
       "2             Spatial transcriptomics in mouse: Puck_191109_20   \n",
       "3             Spatial transcriptomics in mouse: Puck_191112_13   \n",
       "4                                 HTAPP-330-SMP-1082 scRNA-seq   \n",
       "...                                                        ...   \n",
       "1568                              Major cell cluster: Mesoderm   \n",
       "1569                                Tabula Sapiens - All Cells   \n",
       "1570                            Whole dataset: Raw counts only   \n",
       "1571         Whole Taxonomy - MTG: Seattle Alzheimer's Dise...   \n",
       "1572         Whole Taxonomy - DLPFC: Seattle Alzheimer's Di...   \n",
       "\n",
       "                                     dataset_h5ad_path  \\\n",
       "soma_joinid                                              \n",
       "0            4eb29386-de81-452f-b3c0-e00844e8c7fd.h5ad   \n",
       "1            78d59e4a-82eb-4a61-a1dc-da974d7ea54b.h5ad   \n",
       "2            add5eb84-5fc9-4f01-982e-a346dd42ee82.h5ad   \n",
       "3            b020294c-ab82-4547-b5a7-63d8ffa575ed.h5ad   \n",
       "4            d7476ae2-e320-4703-8304-da5c42627e71.h5ad   \n",
       "...                                                ...   \n",
       "1568         dcfa2614-7ca7-4d82-814c-350626eccb26.h5ad   \n",
       "1569         53d208b0-2cfd-4366-9866-c3c6114081bc.h5ad   \n",
       "1570         dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3.h5ad   \n",
       "1571         c2876b1b-06d8-4d96-a56b-5304f815b99a.h5ad   \n",
       "1572         6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3.h5ad   \n",
       "\n",
       "             dataset_total_cell_count  \n",
       "soma_joinid                            \n",
       "0                               10888  \n",
       "1                               10250  \n",
       "2                               12906  \n",
       "3                               15161  \n",
       "4                                 565  \n",
       "...                               ...  \n",
       "1568                          3267338  \n",
       "1569                          1136218  \n",
       "1570                         11441407  \n",
       "1571                          1226855  \n",
       "1572                          1309414  \n",
       "\n",
       "[1573 rows x 10 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import cellxgene_census\n",
    "\n",
    "census = cellxgene_census.open_soma(census_version=\"2025-01-30\")\n",
    "census_datasets = census[\"census_info\"][\"datasets\"].read().concat().to_pandas()\n",
    "\n",
    "# for convenience, indexing on the soma_joinid which links this to other census data.\n",
    "census_datasets = census_datasets.set_index(\"soma_joinid\")\n",
    "\n",
    "census_datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a8bc3c6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_description_prompt(row):\n",
    "    prompt = f\"\"\"\n",
    "Please summarize the following RNA-seq dataset clearly and concisely in 100–150 words, and also identify the experimental technology used.\n",
    "\n",
    "Focus your summary on:\n",
    "- The tissue or organ studied\n",
    "- Organism\n",
    "- Number of cells and samples (if available)\n",
    "- Sample conditions such as health status, age, sex, or strain\n",
    "- Study purpose and biological context (e.g. disease studied, pathways analyzed)\n",
    "- Major annotated or observed cell types\n",
    "\n",
    "Then identify the **specific sequencing technology** used in the experiment (e.g. Smart-seq2, 10x Genomics Chromium, Slide-seqV2, Drop-seq, Microwell-seq, sci-RNA-seq, etc).\n",
    "Do not guess based on publication year or general assumptions. Instead, carefully read the provided title, citation, and collection description for evidence.\n",
    "Do **not** use uncertain language like \"likely\", \"possibly\", or \"based on\". Only output a specific technology name if the content supports it through wording, phrasing, or indirect description.\n",
    "If the technique is not clearly indicated, leave the field blank.\n",
    "\n",
    "You may use clues from the citation, title, or collection name to infer the likely sequencing technology.\n",
    "For example, a 2015 study using tumor biopsy samples likely used Smart-seq2 or Drop-seq. Studies by 10x Genomics likely used 10x Chromium.\n",
    "\n",
    "Strictly format your response as:\n",
    "Description: <one-paragraph English summary>\n",
    "Technology: <exact technology name only — no extra words, no sentences, no parentheses>\n",
    "\n",
    "Do not include any additional text, explanations, or bullet points.\n",
    "\n",
    "Dataset Title: {row['dataset_title']}\n",
    "Citation: {row['citation']}\n",
    "DOI: {row['collection_doi']} ({row['collection_doi_label']})\n",
    "Collection Name: {row['collection_name']} (ID: {row['collection_id']})\n",
    "Cell Count: {row['dataset_total_cell_count']:,} cells\n",
    "\n",
    "Describe this dataset in natural language:\n",
    "\"\"\".strip()\n",
    "    return prompt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "60cfed54",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "User: \n",
      "Please summarize the following RNA-seq dataset clearly and concisely in 100–150 words, and also identify the experimental technology used.\n",
      "\n",
      "Focus your summary on:\n",
      "- The tissue or organ studied\n",
      "- Organism\n",
      "- Number of cells and samples (if available)\n",
      "- Sample conditions such as health status, age, sex, or strain\n",
      "- Study purpose and biological context (e.g. disease studied, pathways analyzed)\n",
      "- Major annotated or observed cell types\n",
      "\n",
      "Then identify the **specific sequencing technology** used in the experiment (e.g. Smart-seq2, 10x Genomics Chromium, Slide-seqV2, Drop-seq, Microwell-seq, sci-RNA-seq, etc).\n",
      "Do not guess based on publication year or general assumptions. Instead, carefully read the provided title, citation, and collection description for evidence.\n",
      "Do **not** use uncertain language like \"likely\", \"possibly\", or \"based on\". Only output a specific technology name if the content supports it through wording, phrasing, or indirect description.\n",
      "If the technique is not clearly indicated, leave the field blank.\n",
      "\n",
      "You may use clues from the citation, title, or collection name to infer the likely sequencing technology.\n",
      "For example, a 2015 study using tumor biopsy samples likely used Smart-seq2 or Drop-seq. Studies by 10x Genomics likely used 10x Chromium.\n",
      "\n",
      "Strictly format your response as:\n",
      "Description: <one-paragraph English summary>\n",
      "Technology: <exact technology name only — no extra words, no sentences, no parentheses>\n",
      "\n",
      "Do not include any additional text, explanations, or bullet points.\n",
      "\n",
      "Dataset Title: Spatial transcriptomics in mouse: Puck_191112_05\n",
      "Citation: Publication: https://doi.org/10.1016/j.isci.2022.104097 Dataset Version: https://datasets.cellxgene.cziscience.com/f76861bb-becb-4eb7-82fc-782dc96ccc7f.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/8e880741-bf9a-4c8e-9227-934204631d2a\n",
      "DOI: 10.1016/j.isci.2022.104097 (Marshall et al. (2022) iScience)\n",
      "Collection Name: High Resolution Slide-seqV2 Spatial Transcriptomics Enables Discovery of Disease-Specific Cell Neighborhoods and Pathways (ID: 8e880741-bf9a-4c8e-9227-934204631d2a)\n",
      "Cell Count: 10,888 cells\n",
      "\n",
      "Describe this dataset in natural language:\n",
      "\n",
      "Assistant (deepseek/deepseek-r1-0528:free): \n",
      "Description: This dataset contains spatial transcriptomics data from one mouse tissue sample (Puck_191112_05), comprising 10,888 cells. The study investigated disease-specific cellular neighborhoods and molecular pathways, implying a disease context, but specific tissue type, disease model, health status, age, sex, strain, or annotated cell types are not explicitly detailed in the provided metadata. Biological focus centers on spatial organization and disease mechanisms.\n",
      "\n",
      "Technology: Slide-seqV2\n"
     ]
    }
   ],
   "source": [
    "import requests, os, json\n",
    "\n",
    "def openrouter_chat(model: str, prompt: str):\n",
    "    resp = requests.post(\n",
    "        \"https://openrouter.ai/api/v1/chat/completions\",\n",
    "        headers={\n",
    "            \"Authorization\": f\"Bearer {os.getenv('OPENROUTER_API_KEY')}\",\n",
    "            \"Content-Type\": \"application/json\"\n",
    "        },\n",
    "        data=json.dumps({\n",
    "            \"model\": model,\n",
    "            \"messages\": [{\"role\": \"user\", \"content\": prompt}]\n",
    "        })\n",
    "    )\n",
    "    return resp.json()\n",
    "\n",
    "# Example usage of the OpenRouter API to summarize a dataset description\n",
    "model = \"deepseek/deepseek-r1-0528:free\"\n",
    "query = format_description_prompt(census_datasets.iloc[0])\n",
    "print(\"User: \")\n",
    "print(query)\n",
    "\n",
    "result = openrouter_chat(model, query)\n",
    "try:\n",
    "    message = result[\"choices\"][0][\"message\"][\"content\"]\n",
    "    print(f\"\\nAssistant ({model}): \")\n",
    "    print(message)\n",
    "except KeyError:\n",
    "    print(\"Error in response format:\", result)\n",
    "    # raise ValueError(\"Unexpected response format from OpenRouter API\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d2cd9914",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1007/2837676445.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['raw_response'] = df.apply(generate_description, axis=1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_title</th>\n",
       "      <th>collection_doi</th>\n",
       "      <th>dataset_total_cell_count</th>\n",
       "      <th>description</th>\n",
       "      <th>technology</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>soma_joinid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191112_05</td>\n",
       "      <td>10.1016/j.isci.2022.104097</td>\n",
       "      <td>10888</td>\n",
       "      <td>This dataset characterizes the transcriptome o...</td>\n",
       "      <td>Slide-seqV2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191112_08</td>\n",
       "      <td>10.1016/j.isci.2022.104097</td>\n",
       "      <td>10250</td>\n",
       "      <td>This dataset provides spatial transcriptomics ...</td>\n",
       "      <td>Slide-seqV2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191109_20</td>\n",
       "      <td>10.1016/j.isci.2022.104097</td>\n",
       "      <td>12906</td>\n",
       "      <td>This dataset captures gene expression patterns...</td>\n",
       "      <td>Slide-seqV2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191112_13</td>\n",
       "      <td>10.1016/j.isci.2022.104097</td>\n",
       "      <td>15161</td>\n",
       "      <td>This dataset represents a high-resolution spat...</td>\n",
       "      <td>Slide-seqV2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HTAPP-330-SMP-1082 scRNA-seq</td>\n",
       "      <td>10.1038/s41591-024-03215-z</td>\n",
       "      <td>565</td>\n",
       "      <td>This dataset includes single-cell RNA sequenci...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191109_14</td>\n",
       "      <td>10.1016/j.isci.2022.104097</td>\n",
       "      <td>12351</td>\n",
       "      <td>This dataset offers high-resolution spatial tr...</td>\n",
       "      <td>Slide-seqV2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Healthy human liver: B cells</td>\n",
       "      <td>10.1002/hep4.1854</td>\n",
       "      <td>146</td>\n",
       "      <td>This dataset comprises single-nucleus RNA sequ...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>thymus scRNA-seq atlas - myeloid p2 subset</td>\n",
       "      <td>10.1038/s41586-024-07944-6</td>\n",
       "      <td>843</td>\n",
       "      <td>This dataset explores the human thymus, a cent...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191109_18</td>\n",
       "      <td>10.1016/j.isci.2022.104097</td>\n",
       "      <td>19156</td>\n",
       "      <td>This dataset explores spatial gene expression ...</td>\n",
       "      <td>Slide-seqV2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Human tonsil nonlymphoid cells scRNA</td>\n",
       "      <td>10.1126/sciimmunol.abe6291</td>\n",
       "      <td>363</td>\n",
       "      <td>This single-cell RNA sequencing dataset invest...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                dataset_title  \\\n",
       "soma_joinid                                                     \n",
       "0            Spatial transcriptomics in mouse: Puck_191112_05   \n",
       "1            Spatial transcriptomics in mouse: Puck_191112_08   \n",
       "2            Spatial transcriptomics in mouse: Puck_191109_20   \n",
       "3            Spatial transcriptomics in mouse: Puck_191112_13   \n",
       "4                                HTAPP-330-SMP-1082 scRNA-seq   \n",
       "5            Spatial transcriptomics in mouse: Puck_191109_14   \n",
       "6                                Healthy human liver: B cells   \n",
       "7                  thymus scRNA-seq atlas - myeloid p2 subset   \n",
       "8            Spatial transcriptomics in mouse: Puck_191109_18   \n",
       "9                        Human tonsil nonlymphoid cells scRNA   \n",
       "\n",
       "                         collection_doi  dataset_total_cell_count  \\\n",
       "soma_joinid                                                         \n",
       "0            10.1016/j.isci.2022.104097                     10888   \n",
       "1            10.1016/j.isci.2022.104097                     10250   \n",
       "2            10.1016/j.isci.2022.104097                     12906   \n",
       "3            10.1016/j.isci.2022.104097                     15161   \n",
       "4            10.1038/s41591-024-03215-z                       565   \n",
       "5            10.1016/j.isci.2022.104097                     12351   \n",
       "6                     10.1002/hep4.1854                       146   \n",
       "7            10.1038/s41586-024-07944-6                       843   \n",
       "8            10.1016/j.isci.2022.104097                     19156   \n",
       "9            10.1126/sciimmunol.abe6291                       363   \n",
       "\n",
       "                                                   description   technology  \n",
       "soma_joinid                                                                  \n",
       "0            This dataset characterizes the transcriptome o...  Slide-seqV2  \n",
       "1            This dataset provides spatial transcriptomics ...  Slide-seqV2  \n",
       "2            This dataset captures gene expression patterns...  Slide-seqV2  \n",
       "3            This dataset represents a high-resolution spat...  Slide-seqV2  \n",
       "4            This dataset includes single-cell RNA sequenci...               \n",
       "5            This dataset offers high-resolution spatial tr...  Slide-seqV2  \n",
       "6            This dataset comprises single-nucleus RNA sequ...               \n",
       "7            This dataset explores the human thymus, a cent...               \n",
       "8            This dataset explores spatial gene expression ...  Slide-seqV2  \n",
       "9            This single-cell RNA sequencing dataset invest...               "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# The census_datasets DataFrame is large, this is a subset for demonstration\n",
    "df = census_datasets[0:10]\n",
    "\n",
    "def generate_description(row, model=\"deepseek/deepseek-r1-0528-qwen3-8b:free\"):\n",
    "    prompt = format_description_prompt(row)\n",
    "    result = openrouter_chat(model, prompt)\n",
    "    try:\n",
    "        return result[\"choices\"][0][\"message\"][\"content\"].strip()\n",
    "    except (KeyError, IndexError) as e:\n",
    "        return f\"[Error retrieving description: {e}]\"\n",
    "\n",
    "df['raw_response'] = df.apply(generate_description, axis=1)\n",
    "\n",
    "def parse_response(text):\n",
    "    if text.startswith(\"[Error\"):\n",
    "        return pd.Series({'description': text, 'technology': \"Error\"})\n",
    "    \n",
    "    parts = text.split(\"Technology:\")\n",
    "    if len(parts) == 2:\n",
    "        description = parts[0].replace(\"Description:\", \"\").strip()\n",
    "        technology = parts[1].strip()\n",
    "    else:\n",
    "        description = text.strip()\n",
    "        technology = \"Unknown\"\n",
    "    return pd.Series({'description': description, 'technology': technology})\n",
    "\n",
    "parsed = df['raw_response'].apply(parse_response)\n",
    "\n",
    "new_df = pd.DataFrame({\n",
    "    'dataset_title': df['dataset_title'],\n",
    "    'collection_doi': df['collection_doi'],\n",
    "    'dataset_total_cell_count': df['dataset_total_cell_count'],\n",
    "    'description': parsed['description'],\n",
    "    'technology': parsed['technology']\n",
    "})\n",
    "\n",
    "new_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4dfcc53d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Description:\\nThis dataset provides spatial transcriptomics data from mouse tissue, comparing normal versus diseased states, focusing on disease-specific cellular neighborhoods, cell states, and pathways involved in pathology. It analyzed approximately 10,250 cells obtained from distinct tissue regions associated with different disease conditions, enabling a detailed examination of the spatio-temporal dynamics of gene expression changes linked to pathological processes. The study aims to identify localized transcriptional programs and interactions relevant to disease mechanisms within the tissue architecture.\\n\\nTechnology: Slide-seqV2'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['raw_response'][1]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cellxgene-env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
