{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cffabf30",
   "metadata": {},
   "source": [
    "Dataset link\n",
    "\n",
    "https://chanzuckerberg.github.io/cellxgene-census/notebooks/api_demo/census_datasets.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b4a5fa39",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python 3.10.18\n"
     ]
    }
   ],
   "source": [
    "!python --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efa6a0b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !pip install cellxgene_census google-generativeai "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "38dcd630",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>citation</th>\n",
       "      <th>collection_id</th>\n",
       "      <th>collection_name</th>\n",
       "      <th>collection_doi</th>\n",
       "      <th>collection_doi_label</th>\n",
       "      <th>dataset_id</th>\n",
       "      <th>dataset_version_id</th>\n",
       "      <th>dataset_title</th>\n",
       "      <th>dataset_h5ad_path</th>\n",
       "      <th>dataset_total_cell_count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>soma_joinid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Publication: https://doi.org/10.1016/j.isci.20...</td>\n",
       "      <td>8e880741-bf9a-4c8e-9227-934204631d2a</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>10.1016/j.isci.2022.104097</td>\n",
       "      <td>Marshall et al. (2022) iScience</td>\n",
       "      <td>4eb29386-de81-452f-b3c0-e00844e8c7fd</td>\n",
       "      <td>f76861bb-becb-4eb7-82fc-782dc96ccc7f</td>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191112_05</td>\n",
       "      <td>4eb29386-de81-452f-b3c0-e00844e8c7fd.h5ad</td>\n",
       "      <td>10888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Publication: https://doi.org/10.1016/j.isci.20...</td>\n",
       "      <td>8e880741-bf9a-4c8e-9227-934204631d2a</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>10.1016/j.isci.2022.104097</td>\n",
       "      <td>Marshall et al. (2022) iScience</td>\n",
       "      <td>78d59e4a-82eb-4a61-a1dc-da974d7ea54b</td>\n",
       "      <td>7d7ec1b6-6e3f-4aaa-9442-4b22f3424396</td>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191112_08</td>\n",
       "      <td>78d59e4a-82eb-4a61-a1dc-da974d7ea54b.h5ad</td>\n",
       "      <td>10250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Publication: https://doi.org/10.1016/j.isci.20...</td>\n",
       "      <td>8e880741-bf9a-4c8e-9227-934204631d2a</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>10.1016/j.isci.2022.104097</td>\n",
       "      <td>Marshall et al. (2022) iScience</td>\n",
       "      <td>add5eb84-5fc9-4f01-982e-a346dd42ee82</td>\n",
       "      <td>de54aed8-4f73-48f6-9229-418a840e2d82</td>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191109_20</td>\n",
       "      <td>add5eb84-5fc9-4f01-982e-a346dd42ee82.h5ad</td>\n",
       "      <td>12906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Publication: https://doi.org/10.1016/j.isci.20...</td>\n",
       "      <td>8e880741-bf9a-4c8e-9227-934204631d2a</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>10.1016/j.isci.2022.104097</td>\n",
       "      <td>Marshall et al. (2022) iScience</td>\n",
       "      <td>b020294c-ab82-4547-b5a7-63d8ffa575ed</td>\n",
       "      <td>abe4fce1-0859-4a56-ad1e-734d79f0e6c8</td>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191112_13</td>\n",
       "      <td>b020294c-ab82-4547-b5a7-63d8ffa575ed.h5ad</td>\n",
       "      <td>15161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Publication: https://doi.org/10.1038/s41591-02...</td>\n",
       "      <td>a96133de-e951-4e2d-ace6-59db8b3bfb1d</td>\n",
       "      <td>HTAN/HTAPP Broad - Spatio-molecular dissection...</td>\n",
       "      <td>10.1038/s41591-024-03215-z</td>\n",
       "      <td>Klughammer et al. (2024) Nat Med</td>\n",
       "      <td>d7476ae2-e320-4703-8304-da5c42627e71</td>\n",
       "      <td>863fc5e4-bd4a-4681-9c3d-0ee7ef54e327</td>\n",
       "      <td>HTAPP-330-SMP-1082 scRNA-seq</td>\n",
       "      <td>d7476ae2-e320-4703-8304-da5c42627e71.h5ad</td>\n",
       "      <td>565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1568</th>\n",
       "      <td>Publication: https://doi.org/10.1038/s41586-02...</td>\n",
       "      <td>45d5d2c3-bc28-4814-aed6-0bb6f0e11c82</td>\n",
       "      <td>A single-cell transcriptional timelapse of mou...</td>\n",
       "      <td>10.1038/s41586-024-07069-w</td>\n",
       "      <td>Qiu et al. (2024) Nature</td>\n",
       "      <td>dcfa2614-7ca7-4d82-814c-350626eccb26</td>\n",
       "      <td>3002a659-a1a9-4406-9976-99e658e1fbb5</td>\n",
       "      <td>Major cell cluster: Mesoderm</td>\n",
       "      <td>dcfa2614-7ca7-4d82-814c-350626eccb26.h5ad</td>\n",
       "      <td>3267338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1569</th>\n",
       "      <td>Publication: https://doi.org/10.1126/science.a...</td>\n",
       "      <td>e5f58829-1a66-40b5-a624-9046778e74f5</td>\n",
       "      <td>Tabula Sapiens</td>\n",
       "      <td>10.1126/science.abl4896</td>\n",
       "      <td>The Tabula Sapiens Consortium* et al. (2022) S...</td>\n",
       "      <td>53d208b0-2cfd-4366-9866-c3c6114081bc</td>\n",
       "      <td>10df7690-6d10-4029-a47e-0f071bb2df83</td>\n",
       "      <td>Tabula Sapiens - All Cells</td>\n",
       "      <td>53d208b0-2cfd-4366-9866-c3c6114081bc.h5ad</td>\n",
       "      <td>1136218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1570</th>\n",
       "      <td>Publication: https://doi.org/10.1038/s41586-02...</td>\n",
       "      <td>45d5d2c3-bc28-4814-aed6-0bb6f0e11c82</td>\n",
       "      <td>A single-cell transcriptional timelapse of mou...</td>\n",
       "      <td>10.1038/s41586-024-07069-w</td>\n",
       "      <td>Qiu et al. (2024) Nature</td>\n",
       "      <td>dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3</td>\n",
       "      <td>3817734b-0f82-433b-8c38-55b214200fff</td>\n",
       "      <td>Whole dataset: Raw counts only</td>\n",
       "      <td>dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3.h5ad</td>\n",
       "      <td>11441407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1571</th>\n",
       "      <td>Publication: https://doi.org/10.1038/s41593-02...</td>\n",
       "      <td>1ca90a2d-2943-483d-b678-b809bf464c30</td>\n",
       "      <td>SEA-AD: Seattle Alzheimer’s Disease Brain Cell...</td>\n",
       "      <td>10.1038/s41593-024-01774-5</td>\n",
       "      <td>Gabitto et al. (2024) Nat Neurosci</td>\n",
       "      <td>c2876b1b-06d8-4d96-a56b-5304f815b99a</td>\n",
       "      <td>c32964d2-3339-441f-8e56-7177234c7876</td>\n",
       "      <td>Whole Taxonomy - MTG: Seattle Alzheimer's Dise...</td>\n",
       "      <td>c2876b1b-06d8-4d96-a56b-5304f815b99a.h5ad</td>\n",
       "      <td>1226855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1572</th>\n",
       "      <td>Publication: https://doi.org/10.1038/s41593-02...</td>\n",
       "      <td>1ca90a2d-2943-483d-b678-b809bf464c30</td>\n",
       "      <td>SEA-AD: Seattle Alzheimer’s Disease Brain Cell...</td>\n",
       "      <td>10.1038/s41593-024-01774-5</td>\n",
       "      <td>Gabitto et al. (2024) Nat Neurosci</td>\n",
       "      <td>6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3</td>\n",
       "      <td>d3427e8c-c55d-4d4e-b15b-1a8774cd3a4b</td>\n",
       "      <td>Whole Taxonomy - DLPFC: Seattle Alzheimer's Di...</td>\n",
       "      <td>6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3.h5ad</td>\n",
       "      <td>1309414</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1573 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      citation  \\\n",
       "soma_joinid                                                      \n",
       "0            Publication: https://doi.org/10.1016/j.isci.20...   \n",
       "1            Publication: https://doi.org/10.1016/j.isci.20...   \n",
       "2            Publication: https://doi.org/10.1016/j.isci.20...   \n",
       "3            Publication: https://doi.org/10.1016/j.isci.20...   \n",
       "4            Publication: https://doi.org/10.1038/s41591-02...   \n",
       "...                                                        ...   \n",
       "1568         Publication: https://doi.org/10.1038/s41586-02...   \n",
       "1569         Publication: https://doi.org/10.1126/science.a...   \n",
       "1570         Publication: https://doi.org/10.1038/s41586-02...   \n",
       "1571         Publication: https://doi.org/10.1038/s41593-02...   \n",
       "1572         Publication: https://doi.org/10.1038/s41593-02...   \n",
       "\n",
       "                                    collection_id  \\\n",
       "soma_joinid                                         \n",
       "0            8e880741-bf9a-4c8e-9227-934204631d2a   \n",
       "1            8e880741-bf9a-4c8e-9227-934204631d2a   \n",
       "2            8e880741-bf9a-4c8e-9227-934204631d2a   \n",
       "3            8e880741-bf9a-4c8e-9227-934204631d2a   \n",
       "4            a96133de-e951-4e2d-ace6-59db8b3bfb1d   \n",
       "...                                           ...   \n",
       "1568         45d5d2c3-bc28-4814-aed6-0bb6f0e11c82   \n",
       "1569         e5f58829-1a66-40b5-a624-9046778e74f5   \n",
       "1570         45d5d2c3-bc28-4814-aed6-0bb6f0e11c82   \n",
       "1571         1ca90a2d-2943-483d-b678-b809bf464c30   \n",
       "1572         1ca90a2d-2943-483d-b678-b809bf464c30   \n",
       "\n",
       "                                               collection_name  \\\n",
       "soma_joinid                                                      \n",
       "0            High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "1            High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "2            High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "3            High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "4            HTAN/HTAPP Broad - Spatio-molecular dissection...   \n",
       "...                                                        ...   \n",
       "1568         A single-cell transcriptional timelapse of mou...   \n",
       "1569                                            Tabula Sapiens   \n",
       "1570         A single-cell transcriptional timelapse of mou...   \n",
       "1571         SEA-AD: Seattle Alzheimer’s Disease Brain Cell...   \n",
       "1572         SEA-AD: Seattle Alzheimer’s Disease Brain Cell...   \n",
       "\n",
       "                         collection_doi  \\\n",
       "soma_joinid                               \n",
       "0            10.1016/j.isci.2022.104097   \n",
       "1            10.1016/j.isci.2022.104097   \n",
       "2            10.1016/j.isci.2022.104097   \n",
       "3            10.1016/j.isci.2022.104097   \n",
       "4            10.1038/s41591-024-03215-z   \n",
       "...                                 ...   \n",
       "1568         10.1038/s41586-024-07069-w   \n",
       "1569            10.1126/science.abl4896   \n",
       "1570         10.1038/s41586-024-07069-w   \n",
       "1571         10.1038/s41593-024-01774-5   \n",
       "1572         10.1038/s41593-024-01774-5   \n",
       "\n",
       "                                          collection_doi_label  \\\n",
       "soma_joinid                                                      \n",
       "0                              Marshall et al. (2022) iScience   \n",
       "1                              Marshall et al. (2022) iScience   \n",
       "2                              Marshall et al. (2022) iScience   \n",
       "3                              Marshall et al. (2022) iScience   \n",
       "4                             Klughammer et al. (2024) Nat Med   \n",
       "...                                                        ...   \n",
       "1568                                  Qiu et al. (2024) Nature   \n",
       "1569         The Tabula Sapiens Consortium* et al. (2022) S...   \n",
       "1570                                  Qiu et al. (2024) Nature   \n",
       "1571                        Gabitto et al. (2024) Nat Neurosci   \n",
       "1572                        Gabitto et al. (2024) Nat Neurosci   \n",
       "\n",
       "                                       dataset_id  \\\n",
       "soma_joinid                                         \n",
       "0            4eb29386-de81-452f-b3c0-e00844e8c7fd   \n",
       "1            78d59e4a-82eb-4a61-a1dc-da974d7ea54b   \n",
       "2            add5eb84-5fc9-4f01-982e-a346dd42ee82   \n",
       "3            b020294c-ab82-4547-b5a7-63d8ffa575ed   \n",
       "4            d7476ae2-e320-4703-8304-da5c42627e71   \n",
       "...                                           ...   \n",
       "1568         dcfa2614-7ca7-4d82-814c-350626eccb26   \n",
       "1569         53d208b0-2cfd-4366-9866-c3c6114081bc   \n",
       "1570         dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3   \n",
       "1571         c2876b1b-06d8-4d96-a56b-5304f815b99a   \n",
       "1572         6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3   \n",
       "\n",
       "                               dataset_version_id  \\\n",
       "soma_joinid                                         \n",
       "0            f76861bb-becb-4eb7-82fc-782dc96ccc7f   \n",
       "1            7d7ec1b6-6e3f-4aaa-9442-4b22f3424396   \n",
       "2            de54aed8-4f73-48f6-9229-418a840e2d82   \n",
       "3            abe4fce1-0859-4a56-ad1e-734d79f0e6c8   \n",
       "4            863fc5e4-bd4a-4681-9c3d-0ee7ef54e327   \n",
       "...                                           ...   \n",
       "1568         3002a659-a1a9-4406-9976-99e658e1fbb5   \n",
       "1569         10df7690-6d10-4029-a47e-0f071bb2df83   \n",
       "1570         3817734b-0f82-433b-8c38-55b214200fff   \n",
       "1571         c32964d2-3339-441f-8e56-7177234c7876   \n",
       "1572         d3427e8c-c55d-4d4e-b15b-1a8774cd3a4b   \n",
       "\n",
       "                                                 dataset_title  \\\n",
       "soma_joinid                                                      \n",
       "0             Spatial transcriptomics in mouse: Puck_191112_05   \n",
       "1             Spatial transcriptomics in mouse: Puck_191112_08   \n",
       "2             Spatial transcriptomics in mouse: Puck_191109_20   \n",
       "3             Spatial transcriptomics in mouse: Puck_191112_13   \n",
       "4                                 HTAPP-330-SMP-1082 scRNA-seq   \n",
       "...                                                        ...   \n",
       "1568                              Major cell cluster: Mesoderm   \n",
       "1569                                Tabula Sapiens - All Cells   \n",
       "1570                            Whole dataset: Raw counts only   \n",
       "1571         Whole Taxonomy - MTG: Seattle Alzheimer's Dise...   \n",
       "1572         Whole Taxonomy - DLPFC: Seattle Alzheimer's Di...   \n",
       "\n",
       "                                     dataset_h5ad_path  \\\n",
       "soma_joinid                                              \n",
       "0            4eb29386-de81-452f-b3c0-e00844e8c7fd.h5ad   \n",
       "1            78d59e4a-82eb-4a61-a1dc-da974d7ea54b.h5ad   \n",
       "2            add5eb84-5fc9-4f01-982e-a346dd42ee82.h5ad   \n",
       "3            b020294c-ab82-4547-b5a7-63d8ffa575ed.h5ad   \n",
       "4            d7476ae2-e320-4703-8304-da5c42627e71.h5ad   \n",
       "...                                                ...   \n",
       "1568         dcfa2614-7ca7-4d82-814c-350626eccb26.h5ad   \n",
       "1569         53d208b0-2cfd-4366-9866-c3c6114081bc.h5ad   \n",
       "1570         dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3.h5ad   \n",
       "1571         c2876b1b-06d8-4d96-a56b-5304f815b99a.h5ad   \n",
       "1572         6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3.h5ad   \n",
       "\n",
       "             dataset_total_cell_count  \n",
       "soma_joinid                            \n",
       "0                               10888  \n",
       "1                               10250  \n",
       "2                               12906  \n",
       "3                               15161  \n",
       "4                                 565  \n",
       "...                               ...  \n",
       "1568                          3267338  \n",
       "1569                          1136218  \n",
       "1570                         11441407  \n",
       "1571                          1226855  \n",
       "1572                          1309414  \n",
       "\n",
       "[1573 rows x 10 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import cellxgene_census\n",
    "\n",
    "census = cellxgene_census.open_soma(census_version=\"2025-01-30\")\n",
    "census_datasets = census[\"census_info\"][\"datasets\"].read().concat().to_pandas()\n",
    "\n",
    "# for convenience, indexing on the soma_joinid which links this to other census data.\n",
    "census_datasets = census_datasets.set_index(\"soma_joinid\")\n",
    "\n",
    "census_datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f26478f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_technology_prompt(row):\n",
    "    prompt = f\"\"\"\n",
    "Based on the following information, identify the **specific RNA-seq technology** used in the experiment.\n",
    "Your output must be **technology name only**, without any explanation, uncertainty, or additional text.\n",
    "Do not include use company names, product names, or any other identifiers. (like \"10x Genomics\", \"10x Visium\", or \"Illumina\").\n",
    "\n",
    "Strictly format your response as:\n",
    "<exact technology name — no extra words, no sentences, no parentheses>\n",
    "\n",
    "Dataset Title: {row['dataset_title']}\n",
    "Citation: {row['citation']}\n",
    "Collection Name: {row['collection_name']}\n",
    "\"\"\".strip()\n",
    "    return prompt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "24c351a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # generate by openrouter API\n",
    "\n",
    "# import requests, os, json\n",
    "\n",
    "# def openrouter_chat(model: str, prompt: str):\n",
    "#     resp = requests.post(\n",
    "#         \"https://openrouter.ai/api/v1/chat/completions\",\n",
    "#         headers={\n",
    "#             \"Authorization\": f\"Bearer {os.getenv('OPENROUTER_API_KEY')}\",\n",
    "#             \"Content-Type\": \"application/json\"\n",
    "#         },\n",
    "#         data=json.dumps({\n",
    "#             \"model\": model,\n",
    "#             \"messages\": [{\"role\": \"user\", \"content\": prompt}]\n",
    "#         })\n",
    "#     )\n",
    "#     return resp.json()\n",
    "\n",
    "# def generate_openrouter_description(row, model=\"deepseek/deepseek-r1-0528-qwen3-8b:free\"):\n",
    "#     prompt = format_technology_prompt(row)\n",
    "#     result = openrouter_chat(model, prompt)\n",
    "#     try:\n",
    "#         return result[\"choices\"][0][\"message\"][\"content\"].strip()\n",
    "#     except (KeyError, IndexError) as e:\n",
    "#         return f\"[Error retrieving description: {e}]\"\n",
    "    \n",
    "# # OpenRouter API to summarize a dataset description\n",
    "\n",
    "# model = \"deepseek/deepseek-r1-0528:free\"\n",
    "# query = format_technology_prompt(census_datasets.iloc[4])\n",
    "# print(\"User: \\n\", query)\n",
    "\n",
    "# result = openrouter_chat(model, query)\n",
    "# try:\n",
    "#     message = result[\"choices\"][0][\"message\"][\"content\"]\n",
    "#     print(f\"\\nAssistant ({model}): \\n {message}\")\n",
    "# except KeyError:\n",
    "#     print(\"Error in response format:\", result)\n",
    "#     # raise ValueError(\"Unexpected response format from OpenRouter API\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "84540a6c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "User: \n",
      " Based on the following information, identify the **specific RNA-seq technology** used in the experiment.\n",
      "Your output must be **technology name only**, without any explanation, uncertainty, or additional text.\n",
      "Do not include use company names, product names, or any other identifiers. (like \"10x Genomics\", \"10x Visium\", or \"Illumina\").\n",
      "\n",
      "Strictly format your response as:\n",
      "<exact technology name — no extra words, no sentences, no parentheses>\n",
      "\n",
      "Dataset Title: HTAPP-330-SMP-1082 scRNA-seq\n",
      "Citation: Publication: https://doi.org/10.1038/s41591-024-03215-z Dataset Version: https://datasets.cellxgene.cziscience.com/863fc5e4-bd4a-4681-9c3d-0ee7ef54e327.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/a96133de-e951-4e2d-ace6-59db8b3bfb1d\n",
      "Collection Name: HTAN/HTAPP Broad - Spatio-molecular dissection of the breast cancer metastatic microenvironment\n",
      "\n",
      "Assistant (models/gemini-1.5-flash-latest): \n",
      " scRNA-seq\n",
      "\n"
     ]
    }
   ],
   "source": [
    "model = \"models/gemini-1.5-flash-latest\"\n",
    "\n",
    "import os\n",
    "import google.generativeai as genai\n",
    "genai.configure(api_key=os.getenv(\"GOOGLE_API_KEY\"))\n",
    "genai_model = genai.GenerativeModel(model)\n",
    "\n",
    "# Gemini API to summarize a dataset description\n",
    "query = format_technology_prompt(census_datasets.iloc[4])\n",
    "print(\"User: \\n\", query)\n",
    "\n",
    "result = genai_model.generate_content(query)\n",
    "try:\n",
    "    print(f\"\\nAssistant ({model}): \\n {result.text}\")\n",
    "except Exception:\n",
    "    print(\"Error in response format:\", result)\n",
    "    # raise ValueError(\"Unexpected response format from OpenRouter API\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "d2cd9914",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pandas as pd\n",
    "\n",
    "# # The census_datasets DataFrame is large, this is a subset for demonstration\n",
    "# df = census_datasets[0:10]\n",
    "\n",
    "# def generate_technology(row):\n",
    "#     prompt = format_technology_prompt(row)\n",
    "#     response = genai_model.generate_content(prompt)\n",
    "#     try:\n",
    "#         return response.text.strip()\n",
    "#     except Exception as e:\n",
    "#         print(f\"Error on row: {row['dataset_title']} — {e}\")\n",
    "#         return \"ERROR\"\n",
    "\n",
    "# df['technology'] = df.apply(generate_technology, axis=1)\n",
    "\n",
    "# new_df = pd.DataFrame({\n",
    "#     'dataset_title': df['dataset_title'],\n",
    "#     'collection_doi': df['collection_doi'],\n",
    "#     'dataset_total_cell_count': df['dataset_total_cell_count'],\n",
    "#     'technology': df['technology']\n",
    "# })\n",
    "\n",
    "# new_df.to_csv(\"census_technology.csv\", index=False)\n",
    "# new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1982c4e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded 14 rows from existing file.\n",
      "Row 14 generated: Slide-seqV2\n",
      "Row 15 generated: Slide-seqV2\n",
      "Row 16 generated: Slide-seqV2\n",
      "Row 17 generated: Slide-seqV2\n",
      "Row 18 generated: Spatial Transcriptomics\n",
      "Row 19 generated: Spatial transcriptomics\n",
      "Row 20 generated: Slide-seqV2\n",
      "Row 21 generated: slide-seqV2\n",
      "Row 22 generated: scRNA-seq\n",
      "Row 23 generated: Slide-seqV2\n",
      "Row 24 generated: Single-cell RNA sequencing\n",
      "Row 25 generated: Slide-seqV2\n",
      "Row 26 generated: Spatial transcriptomics\n",
      "Row 27 generated: Slide-seqV2\n",
      "Row 28 generated: Spatial transcriptomics\n",
      "Stopped at row 29 due to error: 429 Resource has been exhausted (e.g. check quota).\n",
      "\n",
      "Saved to: census_technology_updated.csv\n",
      "Generated from index 14 to 28 this run.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "model = \"models/gemini-1.5-flash-latest\"\n",
    "SAVE_FILE = \"census_technology.csv\"\n",
    "UPDATED_FILE = \"census_technology_updated.csv\"\n",
    "df = census_datasets\n",
    "\n",
    "if os.path.exists(UPDATED_FILE):\n",
    "    existing_df = pd.read_csv(UPDATED_FILE)\n",
    "    existing_df.to_csv(SAVE_FILE, index=False)\n",
    "    print(f\"Loaded {len(existing_df)} rows from existing file.\")\n",
    "else:\n",
    "    existing_df = pd.DataFrame(columns=[\n",
    "        'dataset_title', 'collection_name', 'technology'\n",
    "    ])\n",
    "    print(\"Starting fresh.\")\n",
    "\n",
    "\n",
    "import google.generativeai as genai\n",
    "genai.configure(api_key=os.getenv(\"GOOGLE_API_KEY\"))\n",
    "genai_model = genai.GenerativeModel(model)\n",
    "\n",
    "start_idx, end_idx = None, None\n",
    "\n",
    "for idx in range(len(df)):\n",
    "    if (idx < len(existing_df)) and pd.notna(existing_df.loc[idx, \"technology\"]) and existing_df.loc[idx, \"technology\"].strip() not in [\"\", \"ERROR\"]:\n",
    "        continue\n",
    "\n",
    "    try:\n",
    "        prompt = format_technology_prompt(df.loc[idx])\n",
    "        response = genai_model.generate_content(prompt)\n",
    "        tech = response.text.strip()\n",
    "\n",
    "        existing_df.loc[idx] = {\n",
    "            'dataset_title': df.loc[idx, 'dataset_title'],\n",
    "            'collection_name': df.loc[idx, 'collection_name'],\n",
    "            'technology': tech\n",
    "        }\n",
    "\n",
    "        if start_idx is None:\n",
    "            start_idx = idx\n",
    "        end_idx = idx\n",
    "\n",
    "        print(f\"Row {idx} generated: {tech}\")\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"Stopped at row {idx} due to error: {e}\")\n",
    "        break\n",
    "\n",
    "existing_df.to_csv(UPDATED_FILE, index=False)\n",
    "print(f\"\\nSaved to: {UPDATED_FILE}\")\n",
    "\n",
    "if start_idx is not None:\n",
    "    print(f\"Generated from index {start_idx} to {end_idx} this run.\")\n",
    "else:\n",
    "    print(\"No new rows were generated.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "58c887c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_title</th>\n",
       "      <th>collection_name</th>\n",
       "      <th>technology</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191112_05</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>Slide-seqV2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191112_08</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>Slide-seqV2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191109_20</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>Slide-seqV2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191112_13</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>Slide-seqV2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HTAPP-330-SMP-1082 scRNA-seq</td>\n",
       "      <td>HTAN/HTAPP Broad - Spatio-molecular dissection...</td>\n",
       "      <td>scRNA-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191109_14</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>Slide-seqV2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Healthy human liver: B cells</td>\n",
       "      <td>Single-Cell, Single-Nucleus, and Spatial RNA S...</td>\n",
       "      <td>Single-cell RNA sequencing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>thymus scRNA-seq atlas - myeloid p2 subset</td>\n",
       "      <td>A spatial human thymus cell atlas mapped to a ...</td>\n",
       "      <td>scRNA-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191109_18</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>Slide-seqV2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Human tonsil nonlymphoid cells scRNA</td>\n",
       "      <td>Single-cell analysis of human B cell maturatio...</td>\n",
       "      <td>scRNA-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191109_19</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>Slide-seqV2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191223_13</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>Slide-seqV2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191204_18</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>Slide-seqV2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>mouse_trunkTbx6WT_KO</td>\n",
       "      <td>Spatial transcriptomic maps of whole mouse emb...</td>\n",
       "      <td>Spatial Transcriptomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191109_10</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>Slide-seqV2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191112_07</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>Slide-seqV2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191112_04</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>Slide-seqV2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_200104_04</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>Slide-seqV2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Molecular characterization of selectively vuln...</td>\n",
       "      <td>Molecular characterization of selectively vuln...</td>\n",
       "      <td>Spatial Transcriptomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>E9.0: Puck_220718_02</td>\n",
       "      <td>Spatial transcriptomic maps of whole mouse emb...</td>\n",
       "      <td>Spatial transcriptomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Spatial transcriptomics in the healthy human k...</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>Slide-seqV2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>slide-seqV2 analysis of aorta</td>\n",
       "      <td>A Coding and Non-Coding Atlas of the Human Art...</td>\n",
       "      <td>slide-seqV2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Steady-state B cells - scRNA-seq</td>\n",
       "      <td>Single-cell Atlas of common variable immunodef...</td>\n",
       "      <td>scRNA-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191109_08</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>Slide-seqV2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>blood and bone marrow from a healthy young donor</td>\n",
       "      <td>Single-cell proteo-genomic reference maps of t...</td>\n",
       "      <td>Single-cell RNA sequencing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191204_09</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>Slide-seqV2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Normal Spleen OT1 Puck_200727_02</td>\n",
       "      <td>Spatially mapping T cell receptors and transcr...</td>\n",
       "      <td>Spatial transcriptomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_200104_03</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>Slide-seqV2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>E9.0: Puck_220718_10</td>\n",
       "      <td>Spatial transcriptomic maps of whole mouse emb...</td>\n",
       "      <td>Spatial transcriptomics</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        dataset_title  \\\n",
       "0    Spatial transcriptomics in mouse: Puck_191112_05   \n",
       "1    Spatial transcriptomics in mouse: Puck_191112_08   \n",
       "2    Spatial transcriptomics in mouse: Puck_191109_20   \n",
       "3    Spatial transcriptomics in mouse: Puck_191112_13   \n",
       "4                        HTAPP-330-SMP-1082 scRNA-seq   \n",
       "5    Spatial transcriptomics in mouse: Puck_191109_14   \n",
       "6                        Healthy human liver: B cells   \n",
       "7          thymus scRNA-seq atlas - myeloid p2 subset   \n",
       "8    Spatial transcriptomics in mouse: Puck_191109_18   \n",
       "9                Human tonsil nonlymphoid cells scRNA   \n",
       "10   Spatial transcriptomics in mouse: Puck_191109_19   \n",
       "11   Spatial transcriptomics in mouse: Puck_191223_13   \n",
       "12   Spatial transcriptomics in mouse: Puck_191204_18   \n",
       "13                               mouse_trunkTbx6WT_KO   \n",
       "14   Spatial transcriptomics in mouse: Puck_191109_10   \n",
       "15   Spatial transcriptomics in mouse: Puck_191112_07   \n",
       "16   Spatial transcriptomics in mouse: Puck_191112_04   \n",
       "17   Spatial transcriptomics in mouse: Puck_200104_04   \n",
       "18  Molecular characterization of selectively vuln...   \n",
       "19                               E9.0: Puck_220718_02   \n",
       "20  Spatial transcriptomics in the healthy human k...   \n",
       "21                      slide-seqV2 analysis of aorta   \n",
       "22                   Steady-state B cells - scRNA-seq   \n",
       "23   Spatial transcriptomics in mouse: Puck_191109_08   \n",
       "24   blood and bone marrow from a healthy young donor   \n",
       "25   Spatial transcriptomics in mouse: Puck_191204_09   \n",
       "26                   Normal Spleen OT1 Puck_200727_02   \n",
       "27   Spatial transcriptomics in mouse: Puck_200104_03   \n",
       "28                               E9.0: Puck_220718_10   \n",
       "\n",
       "                                      collection_name  \\\n",
       "0   High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "1   High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "2   High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "3   High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "4   HTAN/HTAPP Broad - Spatio-molecular dissection...   \n",
       "5   High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "6   Single-Cell, Single-Nucleus, and Spatial RNA S...   \n",
       "7   A spatial human thymus cell atlas mapped to a ...   \n",
       "8   High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "9   Single-cell analysis of human B cell maturatio...   \n",
       "10  High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "11  High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "12  High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "13  Spatial transcriptomic maps of whole mouse emb...   \n",
       "14  High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "15  High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "16  High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "17  High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "18  Molecular characterization of selectively vuln...   \n",
       "19  Spatial transcriptomic maps of whole mouse emb...   \n",
       "20  High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "21  A Coding and Non-Coding Atlas of the Human Art...   \n",
       "22  Single-cell Atlas of common variable immunodef...   \n",
       "23  High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "24  Single-cell proteo-genomic reference maps of t...   \n",
       "25  High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "26  Spatially mapping T cell receptors and transcr...   \n",
       "27  High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "28  Spatial transcriptomic maps of whole mouse emb...   \n",
       "\n",
       "                    technology  \n",
       "0                  Slide-seqV2  \n",
       "1                  Slide-seqV2  \n",
       "2                  Slide-seqV2  \n",
       "3                  Slide-seqV2  \n",
       "4                    scRNA-seq  \n",
       "5                  Slide-seqV2  \n",
       "6   Single-cell RNA sequencing  \n",
       "7                    scRNA-seq  \n",
       "8                  Slide-seqV2  \n",
       "9                    scRNA-seq  \n",
       "10                 Slide-seqV2  \n",
       "11                 Slide-seqV2  \n",
       "12                 Slide-seqV2  \n",
       "13     Spatial Transcriptomics  \n",
       "14                 Slide-seqV2  \n",
       "15                 Slide-seqV2  \n",
       "16                 Slide-seqV2  \n",
       "17                 Slide-seqV2  \n",
       "18     Spatial Transcriptomics  \n",
       "19     Spatial transcriptomics  \n",
       "20                 Slide-seqV2  \n",
       "21                 slide-seqV2  \n",
       "22                   scRNA-seq  \n",
       "23                 Slide-seqV2  \n",
       "24  Single-cell RNA sequencing  \n",
       "25                 Slide-seqV2  \n",
       "26     Spatial transcriptomics  \n",
       "27                 Slide-seqV2  \n",
       "28     Spatial transcriptomics  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "existing_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cellxgene-env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
