{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d22133ce",
   "metadata": {},
   "source": [
    "Dataset link\n",
    "\n",
    "https://chanzuckerberg.github.io/cellxgene-census/notebooks/api_demo/census_datasets.html\n",
    "\n",
    "Pub Med link\n",
    "\n",
    "https://pubmed.ncbi.nlm.nih.gov/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "67fbacf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Constants\n",
    "OUTPUT_CSV = \"methods_sections.csv\"\n",
    "BATCH_SLEEP = 1.5  # to avoid getting blocked\n",
    "EMAIL = \"test@example.com\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8352a154",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>citation</th>\n",
       "      <th>collection_id</th>\n",
       "      <th>collection_name</th>\n",
       "      <th>collection_doi</th>\n",
       "      <th>collection_doi_label</th>\n",
       "      <th>dataset_id</th>\n",
       "      <th>dataset_version_id</th>\n",
       "      <th>dataset_title</th>\n",
       "      <th>dataset_h5ad_path</th>\n",
       "      <th>dataset_total_cell_count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>soma_joinid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Publication: https://doi.org/10.1016/j.isci.20...</td>\n",
       "      <td>8e880741-bf9a-4c8e-9227-934204631d2a</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>10.1016/j.isci.2022.104097</td>\n",
       "      <td>Marshall et al. (2022) iScience</td>\n",
       "      <td>4eb29386-de81-452f-b3c0-e00844e8c7fd</td>\n",
       "      <td>f76861bb-becb-4eb7-82fc-782dc96ccc7f</td>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191112_05</td>\n",
       "      <td>4eb29386-de81-452f-b3c0-e00844e8c7fd.h5ad</td>\n",
       "      <td>10888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Publication: https://doi.org/10.1016/j.isci.20...</td>\n",
       "      <td>8e880741-bf9a-4c8e-9227-934204631d2a</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>10.1016/j.isci.2022.104097</td>\n",
       "      <td>Marshall et al. (2022) iScience</td>\n",
       "      <td>78d59e4a-82eb-4a61-a1dc-da974d7ea54b</td>\n",
       "      <td>7d7ec1b6-6e3f-4aaa-9442-4b22f3424396</td>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191112_08</td>\n",
       "      <td>78d59e4a-82eb-4a61-a1dc-da974d7ea54b.h5ad</td>\n",
       "      <td>10250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Publication: https://doi.org/10.1016/j.isci.20...</td>\n",
       "      <td>8e880741-bf9a-4c8e-9227-934204631d2a</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>10.1016/j.isci.2022.104097</td>\n",
       "      <td>Marshall et al. (2022) iScience</td>\n",
       "      <td>add5eb84-5fc9-4f01-982e-a346dd42ee82</td>\n",
       "      <td>de54aed8-4f73-48f6-9229-418a840e2d82</td>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191109_20</td>\n",
       "      <td>add5eb84-5fc9-4f01-982e-a346dd42ee82.h5ad</td>\n",
       "      <td>12906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Publication: https://doi.org/10.1016/j.isci.20...</td>\n",
       "      <td>8e880741-bf9a-4c8e-9227-934204631d2a</td>\n",
       "      <td>High Resolution Slide-seqV2 Spatial Transcript...</td>\n",
       "      <td>10.1016/j.isci.2022.104097</td>\n",
       "      <td>Marshall et al. (2022) iScience</td>\n",
       "      <td>b020294c-ab82-4547-b5a7-63d8ffa575ed</td>\n",
       "      <td>abe4fce1-0859-4a56-ad1e-734d79f0e6c8</td>\n",
       "      <td>Spatial transcriptomics in mouse: Puck_191112_13</td>\n",
       "      <td>b020294c-ab82-4547-b5a7-63d8ffa575ed.h5ad</td>\n",
       "      <td>15161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Publication: https://doi.org/10.1038/s41591-02...</td>\n",
       "      <td>a96133de-e951-4e2d-ace6-59db8b3bfb1d</td>\n",
       "      <td>HTAN/HTAPP Broad - Spatio-molecular dissection...</td>\n",
       "      <td>10.1038/s41591-024-03215-z</td>\n",
       "      <td>Klughammer et al. (2024) Nat Med</td>\n",
       "      <td>d7476ae2-e320-4703-8304-da5c42627e71</td>\n",
       "      <td>863fc5e4-bd4a-4681-9c3d-0ee7ef54e327</td>\n",
       "      <td>HTAPP-330-SMP-1082 scRNA-seq</td>\n",
       "      <td>d7476ae2-e320-4703-8304-da5c42627e71.h5ad</td>\n",
       "      <td>565</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      citation  \\\n",
       "soma_joinid                                                      \n",
       "0            Publication: https://doi.org/10.1016/j.isci.20...   \n",
       "1            Publication: https://doi.org/10.1016/j.isci.20...   \n",
       "2            Publication: https://doi.org/10.1016/j.isci.20...   \n",
       "3            Publication: https://doi.org/10.1016/j.isci.20...   \n",
       "4            Publication: https://doi.org/10.1038/s41591-02...   \n",
       "\n",
       "                                    collection_id  \\\n",
       "soma_joinid                                         \n",
       "0            8e880741-bf9a-4c8e-9227-934204631d2a   \n",
       "1            8e880741-bf9a-4c8e-9227-934204631d2a   \n",
       "2            8e880741-bf9a-4c8e-9227-934204631d2a   \n",
       "3            8e880741-bf9a-4c8e-9227-934204631d2a   \n",
       "4            a96133de-e951-4e2d-ace6-59db8b3bfb1d   \n",
       "\n",
       "                                               collection_name  \\\n",
       "soma_joinid                                                      \n",
       "0            High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "1            High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "2            High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "3            High Resolution Slide-seqV2 Spatial Transcript...   \n",
       "4            HTAN/HTAPP Broad - Spatio-molecular dissection...   \n",
       "\n",
       "                         collection_doi              collection_doi_label  \\\n",
       "soma_joinid                                                                 \n",
       "0            10.1016/j.isci.2022.104097   Marshall et al. (2022) iScience   \n",
       "1            10.1016/j.isci.2022.104097   Marshall et al. (2022) iScience   \n",
       "2            10.1016/j.isci.2022.104097   Marshall et al. (2022) iScience   \n",
       "3            10.1016/j.isci.2022.104097   Marshall et al. (2022) iScience   \n",
       "4            10.1038/s41591-024-03215-z  Klughammer et al. (2024) Nat Med   \n",
       "\n",
       "                                       dataset_id  \\\n",
       "soma_joinid                                         \n",
       "0            4eb29386-de81-452f-b3c0-e00844e8c7fd   \n",
       "1            78d59e4a-82eb-4a61-a1dc-da974d7ea54b   \n",
       "2            add5eb84-5fc9-4f01-982e-a346dd42ee82   \n",
       "3            b020294c-ab82-4547-b5a7-63d8ffa575ed   \n",
       "4            d7476ae2-e320-4703-8304-da5c42627e71   \n",
       "\n",
       "                               dataset_version_id  \\\n",
       "soma_joinid                                         \n",
       "0            f76861bb-becb-4eb7-82fc-782dc96ccc7f   \n",
       "1            7d7ec1b6-6e3f-4aaa-9442-4b22f3424396   \n",
       "2            de54aed8-4f73-48f6-9229-418a840e2d82   \n",
       "3            abe4fce1-0859-4a56-ad1e-734d79f0e6c8   \n",
       "4            863fc5e4-bd4a-4681-9c3d-0ee7ef54e327   \n",
       "\n",
       "                                                dataset_title  \\\n",
       "soma_joinid                                                     \n",
       "0            Spatial transcriptomics in mouse: Puck_191112_05   \n",
       "1            Spatial transcriptomics in mouse: Puck_191112_08   \n",
       "2            Spatial transcriptomics in mouse: Puck_191109_20   \n",
       "3            Spatial transcriptomics in mouse: Puck_191112_13   \n",
       "4                                HTAPP-330-SMP-1082 scRNA-seq   \n",
       "\n",
       "                                     dataset_h5ad_path  \\\n",
       "soma_joinid                                              \n",
       "0            4eb29386-de81-452f-b3c0-e00844e8c7fd.h5ad   \n",
       "1            78d59e4a-82eb-4a61-a1dc-da974d7ea54b.h5ad   \n",
       "2            add5eb84-5fc9-4f01-982e-a346dd42ee82.h5ad   \n",
       "3            b020294c-ab82-4547-b5a7-63d8ffa575ed.h5ad   \n",
       "4            d7476ae2-e320-4703-8304-da5c42627e71.h5ad   \n",
       "\n",
       "             dataset_total_cell_count  \n",
       "soma_joinid                            \n",
       "0                               10888  \n",
       "1                               10250  \n",
       "2                               12906  \n",
       "3                               15161  \n",
       "4                                 565  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import cellxgene_census\n",
    "\n",
    "census = cellxgene_census.open_soma(census_version=\"2025-01-30\")\n",
    "census_datasets = census[\"census_info\"][\"datasets\"].read().concat().to_pandas()\n",
    "\n",
    "# for convenience, indexing on the soma_joinid which links this to other census data.\n",
    "census_datasets = census_datasets.set_index(\"soma_joinid\")\n",
    "\n",
    "census_datasets.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "be671610",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 1558 publication DOIs.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[(0, '10.1016/j.isci.2022.104097'),\n",
       " (1, '10.1016/j.isci.2022.104097'),\n",
       " (2, '10.1016/j.isci.2022.104097'),\n",
       " (3, '10.1016/j.isci.2022.104097'),\n",
       " (4, '10.1038/s41591-024-03215-z'),\n",
       " (5, '10.1016/j.isci.2022.104097'),\n",
       " (6, '10.1002/hep4.1854'),\n",
       " (7, '10.1038/s41586-024-07944-6'),\n",
       " (8, '10.1016/j.isci.2022.104097'),\n",
       " (9, '10.1126/sciimmunol.abe6291')]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import re\n",
    "doi_links = []\n",
    "for i in range(len(census_datasets)):\n",
    "    text = census_datasets.loc[i, \"citation\"]\n",
    "    match = re.search(r'Publication:\\s*https?://doi.org/(\\S+)', text)\n",
    "    if match:\n",
    "        doi_links.append((i, match.group(1)))\n",
    "\n",
    "print(f\"Found {len(doi_links)} publication DOIs.\")\n",
    "doi_links[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b0f4b3fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import requests\n",
    "from bs4 import BeautifulSoup\n",
    "import os\n",
    "import time\n",
    "import json\n",
    "import csv\n",
    "\n",
    "# Constants\n",
    "OUTPUT_CSV = \"methods_sections.csv\"\n",
    "EMAIL = \"test@example.com\"\n",
    "CSV_FIELDS = [\"soma_joinid\", \"doi\", \"title\", \"method_headers\"]\n",
    "if not os.path.exists(OUTPUT_CSV):\n",
    "    with open(OUTPUT_CSV, \"w\", newline=\"\", encoding=\"utf-8\") as f:\n",
    "        writer = csv.DictWriter(f, fieldnames=CSV_FIELDS)\n",
    "        writer.writeheader()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "86251488",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_pmcid_from_doi(doi: str) -> str | None:\n",
    "    url = f\"https://www.ncbi.nlm.nih.gov/pmc/utils/idconv/v1.0/?tool=test&email={EMAIL}&ids={doi}\"\n",
    "    resp = requests.get(url)\n",
    "    # time.sleep(1.5)\n",
    "    soup = BeautifulSoup(resp.text, \"xml\")\n",
    "    return soup.find(\"record\").get(\"pmcid\")\n",
    "\n",
    "def fetch_pmc_xml(pmcid: str) -> BeautifulSoup:\n",
    "    xml_url = f\"https://www.ncbi.nlm.nih.gov/pmc/oai/oai.cgi?verb=GetRecord&identifier=oai:pubmedcentral.nih.gov:{pmcid[3:]}&metadataPrefix=pmc\"\n",
    "    resp = requests.get(xml_url)\n",
    "    time.sleep(BATCH_SLEEP)\n",
    "    return BeautifulSoup(resp.content, \"xml\")\n",
    "\n",
    "def extract_article_title(xml_soup: BeautifulSoup) -> str | None:\n",
    "    # PubMed format\n",
    "    alt_title = xml_soup.find(\"ArticleTitle\")\n",
    "    if alt_title:\n",
    "        return alt_title.get_text(strip=True)\n",
    "    \n",
    "    # PMC full-text format\n",
    "    article_title = xml_soup.find(\"article-title\")\n",
    "    if article_title:\n",
    "        return article_title.get_text(strip=True)\n",
    "\n",
    "    return None\n",
    "\n",
    "def extract_methods_headers_only(xml_soup: BeautifulSoup) -> list[str]:\n",
    "    methods_sec = [\n",
    "        sec for sec in xml_soup.find(\"body\").find_all(\"sec\", recursive=False)\n",
    "        if sec.find(\"title\") and \"method\" in sec.find(\"title\").get_text(strip=True).lower()\n",
    "    ]\n",
    "    if len(methods_sec) != 1:\n",
    "        return []\n",
    "\n",
    "    method = methods_sec[0]\n",
    "    headers = []\n",
    "    for sub_sec in method.find_all(\"sec\", recursive=False):\n",
    "        title = sub_sec.find(\"title\")\n",
    "        if title:\n",
    "            headers.append(title.get_text(strip=True))\n",
    "    return headers\n",
    "\n",
    "def append_headers_to_csv(path: str, records: list[dict], write_header: bool = False):\n",
    "    with open(path, \"a\", newline=\"\", encoding=\"utf-8\") as f:\n",
    "        writer = csv.DictWriter(f, fieldnames=[\"soma_joinid\", \"doi\", \"title\", \"method_headers\"])\n",
    "        if write_header:\n",
    "            writer.writeheader()\n",
    "        for row in records:\n",
    "            row[\"method_headers\"] = json.dumps(row[\"method_headers\"], ensure_ascii=False)\n",
    "            row[\"title\"] = row[\"title\"].strip().replace(\"\\n\", \" \").replace(\"\\r\", \" \")  # clean up line breaks\n",
    "            writer.writerow(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "01646207",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Row 0 - Already processed.\n",
      "Row 1 - Already processed.\n",
      "Row 2 - Already processed.\n",
      "Row 3 - Already processed.\n",
      "Row 4 - Already processed.\n",
      "Row 5 - Already processed.\n",
      "Row 6 - Already processed.\n",
      "Row 7 - Already processed.\n",
      "Row 8 - Already processed.\n",
      "Row 9 - No PMCID for DOI 10.1126/sciimmunol.abe6291\n",
      "Row 10 - Already processed.\n",
      "Row 11 - Already processed.\n",
      "Row 12 - Already processed.\n",
      "Row 13 - Already processed.\n",
      "Row 14 - Already processed.\n",
      "Row 15 - Already processed.\n",
      "Row 16 - Already processed.\n",
      "Row 17 - Already processed.\n",
      "Row 18 - Already processed.\n",
      "Row 19 - Already processed.\n",
      "Row 20 - Already processed.\n",
      "Row 21 - No title found.\n",
      "Row 22 - Already processed.\n",
      "Row 23 - Already processed.\n",
      "Row 24 - Already processed.\n",
      "Row 25 - Already processed.\n",
      "Row 26 - No title found.\n",
      "Row 27 - Already processed.\n",
      "Row 28 - Already processed.\n",
      "Row 29 - Already processed.\n",
      "Row 30 - No title found.\n",
      "Row 31 - No PMCID for DOI 10.1016/j.celrep.2019.12.082\n",
      "Row 32 - Already processed.\n",
      "Row 33 - Already processed.\n",
      "Row 34 - Already processed.\n",
      "Row 35 - Already processed.\n",
      "Row 36 - Already processed.\n",
      "Row 37 - Already processed.\n",
      "Row 38 - Already processed.\n",
      "Row 39 - Already processed.\n",
      "Row 40 - No PMCID for DOI 10.1016/j.cell.2022.11.005\n",
      "Row 41 - Already processed.\n",
      "Row 42 - Already processed.\n",
      "Row 43 - Already processed.\n",
      "Row 44 - Already processed.\n",
      "Row 45 - Already processed.\n",
      "Row 46 - Already processed.\n",
      "Row 47 - Already processed.\n",
      "Row 48 - Already processed.\n",
      "Row 49 - Already processed.\n",
      "Row 50 - Already processed.\n",
      "Row 51 - No PMCID for DOI 10.1016/j.jhep.2023.12.023\n",
      "Row 52 - No title found.\n",
      "Row 53 - Already processed.\n",
      "Row 54 - Already processed.\n",
      "Row 55 - Already processed.\n",
      "Row 56 - No title found.\n",
      "Row 57 - Already processed.\n",
      "Row 58 - Already processed.\n",
      "Row 59 - Already processed.\n",
      "Row 60 - Already processed.\n",
      "Row 61 - Already processed.\n",
      "Row 62 - Already processed.\n",
      "Row 63 - Already processed.\n",
      "Row 64 - Already processed.\n",
      "Row 65 - Already processed.\n",
      "Row 66 - Already processed.\n",
      "Row 67 - Already processed.\n",
      "Row 68 - No PMCID for DOI 10.1016/j.jhep.2023.12.023\n",
      "Row 69 - Already processed.\n",
      "Row 70 - Already processed.\n",
      "Row 71 - Already processed.\n",
      "Row 72 - Already processed.\n",
      "Row 73 - Already processed.\n",
      "Row 74 - Already processed.\n",
      "Row 75 - No title found.\n",
      "Row 76 - Already processed.\n",
      "Row 77 - Already processed.\n",
      "Row 78 - Already processed.\n",
      "Row 79 - Already processed.\n",
      "Row 80 - Already processed.\n",
      "Row 81 - No title found.\n",
      "Row 82 - Already processed.\n",
      "Row 83 - No PMCID for DOI 10.1016/j.jhep.2023.12.023\n",
      "Row 84 - Already processed.\n",
      "Row 85 - Already processed.\n",
      "Row 86 - Already processed.\n",
      "Row 87 - Already processed.\n",
      "Row 88 - Already processed.\n",
      "Row 89 - Already processed.\n",
      "Row 90 - No title found.\n",
      "Row 91 - Already processed.\n",
      "Row 92 - No title found.\n",
      "Row 93 - Already processed.\n",
      "Row 94 - Already processed.\n",
      "Row 95 - Already processed.\n",
      "Row 96 - Already processed.\n",
      "Row 97 - No title found.\n",
      "Row 98 - Already processed.\n",
      "Row 99 - No title found.\n"
     ]
    }
   ],
   "source": [
    "written_df = pd.read_csv(OUTPUT_CSV)\n",
    "written_set = set(zip(written_df[\"soma_joinid\"], written_df[\"doi\"]))\n",
    "\n",
    "# Main processing loop (you can adjust range or remove [:10] limit)\n",
    "for idx, doi in doi_links[:100]:\n",
    "    if (idx, doi) in written_set:\n",
    "        print(f\"Row {idx} - Already processed.\")\n",
    "        continue\n",
    "\n",
    "    try:\n",
    "        pmcid = get_pmcid_from_doi(doi)\n",
    "        if not pmcid:\n",
    "            print(f\"Row {idx} - No PMCID for DOI {doi}\")\n",
    "            continue\n",
    "\n",
    "        xml_soup = fetch_pmc_xml(pmcid)\n",
    "\n",
    "        title = extract_article_title(xml_soup)\n",
    "        if not title:\n",
    "            print(f\"Row {idx} - No title found.\")\n",
    "            continue\n",
    "\n",
    "        headers = extract_methods_headers_only(xml_soup)\n",
    "        if not headers:\n",
    "            print(f\"Row {idx} - No valid Methods section found.\")\n",
    "            continue\n",
    "\n",
    "        record = [{\n",
    "            \"soma_joinid\": idx,\n",
    "            \"doi\": doi,\n",
    "            \"title\": title,\n",
    "            \"method_headers\": headers\n",
    "        }]\n",
    "\n",
    "        append_headers_to_csv(OUTPUT_CSV, record, write_header=False)\n",
    "        print(f\"Row {idx} - Wrote {len(headers)} headers, for '{title[:50]}...'\")\n",
    "    except Exception as e:\n",
    "        print(f\"Row {idx} - Error: {e}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0512d815",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   soma_joinid                         doi  \\\n",
      "0            0  10.1016/j.isci.2022.104097   \n",
      "4            4  10.1038/s41591-024-03215-z   \n",
      "6            6           10.1002/hep4.1854   \n",
      "7            7  10.1038/s41586-024-07944-6   \n",
      "12          13  10.1038/s41588-023-01435-6   \n",
      "\n",
      "                                                title  \\\n",
      "0   High-resolution Slide-seqV2 spatial transcript...   \n",
      "4   A multi-modal single-cell and spatial expressi...   \n",
      "6   Single‐Cell, Single‐Nucleus, and Spatial RNA S...   \n",
      "7   A spatial human thymus cell atlas mapped to a ...   \n",
      "12  Spatiotemporal transcriptomic maps of whole mo...   \n",
      "\n",
      "                                       method_headers  \n",
      "0   [\"Key resources table\", \"Resource availability...  \n",
      "4   [\"Ethics statement\", \"Sample acquisition, hand...  \n",
      "6   [\"Preparation of Fresh Tissue Homogenates and ...  \n",
      "7   [\"Data generation by institute\", \"Sample proce...  \n",
      "12  [\"Animal work and embryo preparation (WT and K...  \n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv(OUTPUT_CSV, dtype={\"soma_joinid\": str, \"doi\": str, \"title\": str, \"method_headers\": str})\n",
    "unique_df = df.drop_duplicates(subset=\"doi\", keep=\"first\")\n",
    "print(unique_df[[\"soma_joinid\", \"doi\", \"title\", \"method_headers\"]].head())\n",
    "unique_df.to_csv(\"unique_method_sections.csv\", index=False, encoding=\"utf-8\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cellxgene-env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
